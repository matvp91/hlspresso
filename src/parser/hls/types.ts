import type { DateTime } from "luxon";
import type {
  HLSDateRange,
  HLSDefine,
  HLSKey,
  HLSMap,
  HLSMediaType,
  HLSPlaylistType,
  HLSResolution,
} from "./lexical-parse";

export type MainPlaylist = {
  independentSegments?: boolean;
  variants: Variant[];
  medias: Media[];
  defines: HLSDefine[];
  comments?: string[];
};

export type Variant = {
  uri: string;
  bandwidth: number;
  codecs?: string;
  resolution?: HLSResolution;
  audio?: string;
  subtitles?: string;
};

export type Media = {
  groupId: string;
  name: string;
  type: HLSMediaType;
  uri: string;
  language?: string;
  channels?: string;
  default?: boolean;
  autoSelect?: boolean;
  characteristics?: string;
};

export type MediaPlaylist = {
  independentSegments?: boolean;
  targetDuration: number;
  endlist: boolean;
  playlistType?: HLSPlaylistType;
  segments: Segment[];
  mediaSequenceBase?: number;
  discontinuitySequenceBase?: number;
  dateRanges: HLSDateRange[];
  defines: HLSDefine[];
};

export type Segment = {
  uri: string;
  duration: number;
  discontinuity?: boolean;
  map?: HLSMap;
  key?: HLSKey;
  programDateTime?: DateTime;
  spliceInfo?: SpliceInfo;
};

export type SpliceInfo = {
  type: "OUT";
  duration?: number;
};
