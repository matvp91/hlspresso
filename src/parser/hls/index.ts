export { parseMainPlaylist, parseMediaPlaylist } from "./parse";

export { stringifyMainPlaylist, stringifyMediaPlaylist } from "./stringify";

export type { MainPlaylist, MediaPlaylist } from "./types";
