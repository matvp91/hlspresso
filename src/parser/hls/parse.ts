import type { DateTime } from "luxon";
import { assert } from "../../assert";
import { lexicalParse, nextLiteral } from "./lexical-parse";
import type {
  HLSDateRange,
  HLSDefine,
  HLSKey,
  HLSMap,
  HLSPlaylistType,
  Tag,
  TagName,
} from "./lexical-parse";
import type {
  MainPlaylist,
  Media,
  MediaPlaylist,
  Segment,
  SpliceInfo,
  Variant,
} from "./types";

export function parseMainPlaylist(text: string): MainPlaylist {
  const tags = lexicalParse(text);

  let independentSegments = false;
  const variants: Variant[] = [];
  const medias: Media[] = [];
  const defines: HLSDefine[] = [];

  tags.forEach(([name, value], index) => {
    if (name === "EXT-X-INDEPENDENT-SEGMENTS") {
      independentSegments = true;
    }
    if (name === "EXT-X-STREAM-INF") {
      const uri = nextLiteral(tags, index);
      variants.push({
        uri,
        bandwidth: value.bandwidth,
        resolution: value.resolution,
        codecs: value.codecs,
        audio: value.audio,
        subtitles: value.subtitles,
      });
    }
    if (name === "EXT-X-MEDIA") {
      medias.push({
        type: value.type,
        groupId: value.groupId,
        name: value.name,
        uri: value.uri,
        channels: value.channels,
        language: value.language,
        default: value.default,
        autoSelect: value.autoSelect,
        characteristics: value.characteristics,
      });
    }
    if (name === "EXT-X-DEFINE") {
      defines.push(value);
    }
  });

  return {
    independentSegments,
    variants,
    medias,
    defines,
  };
}

export function parseMediaPlaylist(text: string): MediaPlaylist {
  const tags = lexicalParse(text);

  let targetDuration: number | undefined;
  let endlist = false;
  let playlistType: HLSPlaylistType | undefined;
  let independentSegments = false;
  let mediaSequenceBase: number | undefined;
  let discontinuitySequenceBase: number | undefined;
  const dateRanges: HLSDateRange[] = [];
  const segments: Segment[] = [];
  const defines: HLSDefine[] = [];

  const splicedDateRanges: HLSDateRange[] = [];

  for (const [name, value] of tags) {
    if (name === "EXT-X-TARGETDURATION") {
      targetDuration = value;
    }
    if (name === "EXT-X-ENDLIST") {
      endlist = true;
    }
    if (name === "EXT-X-PLAYLIST-TYPE") {
      playlistType = value;
    }
    if (name === "EXT-X-INDEPENDENT-SEGMENTS") {
      independentSegments = true;
    }
    if (name === "EXT-X-MEDIA-SEQUENCE") {
      mediaSequenceBase = value;
    }
    if (name === "EXT-X-DISCONTINUITY-SEQUENCE") {
      discontinuitySequenceBase = value;
    }
    if (name === "EXT-X-DATERANGE") {
      if (value.SCTE35) {
        // When the dateRange contains spliced info, we're going to
        // collect it and match it with the corresponding segment.
        splicedDateRanges.push(value);
      } else {
        dateRanges.push(value);
      }
    }
    if (name === "EXT-X-DEFINE") {
      defines.push(value);
    }
  }

  let segmentStart: number | null = null;
  tags.forEach(([name], index) => {
    if (segmentStart === null) {
      if (isSegmentTag(name)) {
        segmentStart = index;
      }
    } else if (name === "LITERAL") {
      const segmentTags = tags.slice(segmentStart, index + 1);
      const segment = parseSegment(segmentTags, splicedDateRanges);
      segments.push(segment);
      segmentStart = null;
    }
  });

  normalizeSegments(segments);

  assert(targetDuration);

  return {
    targetDuration,
    endlist,
    playlistType,
    segments,
    independentSegments,
    mediaSequenceBase,
    discontinuitySequenceBase,
    dateRanges,
    defines,
  };
}

function parseSegment(
  segmentTags: Tag[],
  splicedDateRanges: HLSDateRange[],
): Segment {
  let duration: number | undefined;
  let discontinuity: boolean | undefined;
  let programDateTime: DateTime | undefined;
  let spliceInfo: SpliceInfo | undefined;
  let uri: string | undefined;
  let map: HLSMap | undefined;
  let key: HLSKey | undefined;

  for (const [name, value] of segmentTags) {
    if (name === "EXTINF") {
      duration = value.duration;
    }
    if (name === "EXT-X-DISCONTINUITY") {
      discontinuity = true;
    }
    if (name === "EXT-X-PROGRAM-DATE-TIME") {
      programDateTime = value;
    }
    if (name === "EXT-X-CUE-OUT") {
      spliceInfo = { type: "OUT", duration: value.duration };
    }
    if (name === "EXT-X-MAP") {
      map = value;
    }
    if (name === "EXT-X-KEY") {
      key = value;
    }
    if (name === "LITERAL") {
      uri = value;
    }
  }

  // Find spliceInfo from the dataRange tags.
  if (!spliceInfo && programDateTime) {
    const dateRange = splicedDateRanges.find((item) =>
      item.startDate.equals(programDateTime),
    );
    if (dateRange?.SCTE35) {
      spliceInfo = {
        type: "OUT",
        duration: dateRange.duration,
      };
    }
  }

  assert(uri, "parseSegment: uri not found");
  assert(duration, "parseSegment: duration not found");

  return {
    uri,
    duration,
    map,
    key,
    discontinuity,
    programDateTime,
    spliceInfo,
  };
}

function normalizeSegments(segments: Segment[]) {
  let lastMap: HLSMap | undefined;
  let lastKey: HLSKey | undefined;
  for (const segment of segments) {
    if (segment.map === undefined && lastMap) {
      segment.map = lastMap;
    } else {
      lastMap = segment.map;
    }

    if (segment.key === undefined && lastKey) {
      segment.key = lastKey;
    } else {
      lastKey = segment.key;
    }
  }
}

/**
 * Checks whether the tag name belongs to a segment.
 * @link https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis-16#section-4.4.4
 * @param name
 * @returns
 */
function isSegmentTag(name: TagName) {
  switch (name) {
    case "EXTINF":
    case "EXT-X-DISCONTINUITY":
    case "EXT-X-PROGRAM-DATE-TIME":
    case "EXT-X-MAP":
    case "EXT-X-CUE-OUT":
    case "EXT-X-CUE-IN":
    case "EXT-X-KEY":
      return true;
  }
  return false;
}
