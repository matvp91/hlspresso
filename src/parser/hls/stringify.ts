import { byteSequenceToHex } from "./helpers";
import type { HLSKey, HLSMap } from "./lexical-parse";
import type { MainPlaylist, MediaPlaylist } from "./types";

export function stringifyMainPlaylist(playlist: MainPlaylist) {
  const lines: string[] = [];

  lines.push("#EXTM3U", "#EXT-X-VERSION:8");

  if (playlist.comments) {
    for (const comment of playlist.comments) {
      lines.push(`## ${comment}`);
    }
  }

  if (playlist.independentSegments) {
    lines.push("#EXT-X-INDEPENDENT-SEGMENTS");
  }

  for (const define of playlist.defines) {
    const attrs = [];
    if (define.name !== undefined) {
      attrs.push(`NAME="${define.name}"`);
    }
    if (define.value !== undefined) {
      attrs.push(`VALUE="${define.value}"`);
    }
    if (define.import !== undefined) {
      attrs.push(`IMPORT="${define.import}"`);
    }
    lines.push(`#EXT-X-DEFINE:${attrs.join(",")}`);
  }

  if (playlist.medias.length) {
    lines.push("");
  }

  for (const media of playlist.medias) {
    const attrs = [
      `TYPE=${media.type}`,
      `GROUP-ID="${media.groupId}"`,
      `NAME="${media.name}"`,
    ];
    if (media.language) {
      attrs.push(`LANGUAGE="${media.language}"`);
    }
    if (media.default !== undefined) {
      attrs.push(`DEFAULT=${media.default ? "YES" : "NO"}`);
    }
    if (media.autoSelect !== undefined) {
      attrs.push(`AUTOSELECT=${media.autoSelect ? "YES" : "NO"}`);
    }
    if (media.uri) {
      attrs.push(`URI="${media.uri}"`);
    }
    if (media.channels) {
      attrs.push(`CHANNELS="${media.channels}"`);
    }
    if (media.characteristics) {
      attrs.push(`CHARACTERISTICS="${media.characteristics}"`);
    }
    lines.push(`#EXT-X-MEDIA:${attrs.join(",")}`);
  }

  if (playlist.variants.length) {
    lines.push("");
  }

  for (const variant of playlist.variants) {
    const attrs = [`BANDWIDTH=${variant.bandwidth}`];
    if (variant.codecs) {
      attrs.push(`CODECS="${variant.codecs}"`);
    }
    if (variant.resolution) {
      attrs.push(
        `RESOLUTION=${variant.resolution.width}x${variant.resolution.height}`,
      );
    }
    if (variant.audio) {
      if (
        !playlist.medias.find(
          (media) => media.type === "AUDIO" && media.groupId === variant.audio,
        )
      ) {
        continue;
      }
      attrs.push(`AUDIO="${variant.audio}"`);
    }
    if (variant.subtitles) {
      if (
        !playlist.medias.find(
          (media) =>
            media.type === "SUBTITLES" && media.groupId === variant.subtitles,
        )
      ) {
        continue;
      }
      attrs.push(`SUBTITLES="${variant.subtitles}"`);
    }
    lines.push(`#EXT-X-STREAM-INF:${attrs.join(",")}`);
    lines.push(variant.uri);
  }

  return lines.join("\n");
}

export function stringifyMediaPlaylist(playlist: MediaPlaylist) {
  const lines: string[] = [];

  lines.push(
    "#EXTM3U",
    "#EXT-X-VERSION:8",
    `#EXT-X-TARGETDURATION:${playlist.targetDuration}`,
  );

  if (playlist.playlistType) {
    lines.push(`#EXT-X-PLAYLIST-TYPE:${playlist.playlistType}`);
  }

  for (const define of playlist.defines) {
    const attrs = [];
    if (define.name !== undefined) {
      attrs.push(`NAME="${define.name}"`);
    }
    if (define.value !== undefined) {
      attrs.push(`VALUE="${define.value}"`);
    }
    if (define.import !== undefined) {
      attrs.push(`IMPORT="${define.import}"`);
    }
    lines.push(`#EXT-X-DEFINE:${attrs.join(",")}`);
  }

  if (playlist.independentSegments) {
    lines.push("#EXT-X-INDEPENDENT-SEGMENTS");
  }

  if (playlist.mediaSequenceBase) {
    lines.push(`#EXT-X-MEDIA-SEQUENCE:${playlist.mediaSequenceBase}`);
  }

  if (playlist.discontinuitySequenceBase) {
    lines.push(
      `#EXT-X-DISCONTINUITY-SEQUENCE:${playlist.discontinuitySequenceBase}`,
    );
  }

  let lastMap: HLSMap | undefined;
  let lastKey: HLSKey | undefined;

  for (const segment of playlist.segments) {
    // See https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis-16#section-4.4.4.5
    // It applies to every Media Segment that appears after it in the Playlist until the next
    // EXT-X-MAP tag or until the end of the Playlist.
    if (segment.map !== lastMap) {
      if (segment.map) {
        const attrs = [`URI="${segment.map.uri}"`];
        lines.push(`#EXT-X-MAP:${attrs.join(",")}`);
      }
      lastMap = segment.map;
    }
    if (segment.key !== lastKey) {
      if (segment.key) {
        const attrs = [`METHOD=${segment.key.method}`];
        if (segment.key.uri) {
          attrs.push(`URI="${segment.key.uri}"`);
        }
        if (segment.key.iv) {
          attrs.push(`IV=${byteSequenceToHex(segment.key.iv)}`);
        }
        if (segment.key.format) {
          attrs.push(`KEYFORMAT="${segment.key.format}"`);
        }
        if (segment.key.formatVersion) {
          attrs.push(`KEYFORMATVERSIONS="${segment.key.formatVersion}"`);
        }
        lines.push(`#EXT-X-KEY:${attrs.join(",")}`);
      }
      lastKey = segment.key;
    }

    if (segment.discontinuity) {
      lines.push("#EXT-X-DISCONTINUITY");
    }

    if (segment.spliceInfo) {
      if (segment.spliceInfo.type === "OUT") {
        const attrs = ["DERIVED"];
        if (segment.spliceInfo.duration) {
          attrs.push(`DURATION=${segment.spliceInfo.duration}`);
        }
        lines.push(`#EXT-X-CUE-OUT:${attrs.join(",")}`);
      }
    }

    if (segment.programDateTime) {
      lines.push(`#EXT-X-PROGRAM-DATE-TIME:${segment.programDateTime.toISO()}`);
    }

    let duration = segment.duration.toFixed(3);
    if (duration.match(/\./)) {
      duration = duration.replace(/\.?0+$/, "");
    }
    lines.push(`#EXTINF:${duration}`);

    lines.push(segment.uri);
  }

  if (playlist.endlist) {
    lines.push("#EXT-X-ENDLIST");
  }

  for (const dateRange of playlist.dateRanges) {
    const attrs = [
      `ID="${dateRange.id}"`,
      `START-DATE="${dateRange.startDate.toISO()}"`,
    ];

    if (dateRange.classId) {
      attrs.push(`CLASS="${dateRange.classId}"`);
    }
    if (dateRange.duration) {
      attrs.push(`DURATION=${dateRange.duration}`);
    }
    if (
      dateRange.plannedDuration &&
      dateRange.plannedDuration !== dateRange.duration
    ) {
      attrs.push(`PLANNED-DURATION=${dateRange.plannedDuration}`);
    }

    if (dateRange.custom) {
      const entries = Object.entries(dateRange.custom);
      for (const [key, value] of entries) {
        if (typeof value === "string") {
          attrs.push(`X-${key}="${value}"`);
        }
        if (typeof value === "number") {
          attrs.push(`X-${key}=${value}`);
        }
      }
    }

    lines.push(`#EXT-X-DATERANGE:${attrs.join(",")}`);
  }

  return lines.join("\n");
}
