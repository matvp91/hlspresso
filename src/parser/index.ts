export { parseMainPlaylist, parseMediaPlaylist } from "./parse";

export { stringifyMainPlaylist, stringifyMediaPlaylist } from "./stringify";
