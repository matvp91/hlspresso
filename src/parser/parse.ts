import type { DateTime } from "luxon";
import { assert } from "../assert";
import {
  type HLSDateRange,
  type HLSDefine,
  type HLSKey,
  type HLSMap,
  type HLSPlaylistType,
  type Tag,
  type TagName,
  lexicalParse,
  nextLiteral,
} from "./lexical-parse";
import type {
  MainPlaylist,
  Media,
  MediaPlaylist,
  Segment,
  SpliceInfo,
  Variant,
} from "./types";

export function parseMainPlaylist(text: string): MainPlaylist {
  const tags = lexicalParse(text);

  let independentSegments = false;
  const variants: Variant[] = [];
  const medias: Media[] = [];
  const defines: HLSDefine[] = [];

  tags.forEach(([name, value], index) => {
    if (name === "EXT-X-INDEPENDENT-SEGMENTS") {
      independentSegments = true;
    }
    if (name === "EXT-X-STREAM-INF") {
      const uri = nextLiteral(tags, index);
      variants.push({
        uri,
        bandwidth: value.bandwidth,
        resolution: value.resolution,
        codecs: value.codecs,
        audio: value.audio,
        subtitles: value.subtitles,
      });
    }
    if (name === "EXT-X-MEDIA") {
      medias.push({
        type: value.type,
        groupId: value.groupId,
        name: value.name,
        uri: value.uri,
        channels: value.channels,
        language: value.language,
        default: value.default,
        autoSelect: value.autoSelect,
        characteristics: value.characteristics,
      });
    }
    if (name === "EXT-X-DEFINE") {
      defines.push(value);
    }
  });

  return {
    independentSegments,
    variants,
    medias,
    defines,
  };
}

export function parseMediaPlaylist(text: string): MediaPlaylist {
  const tags = lexicalParse(text);

  let targetDuration: number | undefined;
  let endlist = false;
  let playlistType: HLSPlaylistType | undefined;
  let independentSegments = false;
  let mediaSequenceBase: number | undefined;
  let discontinuitySequenceBase: number | undefined;
  const dateRanges: HLSDateRange[] = [];
  const segments: Segment[] = [];
  const defines: HLSDefine[] = [];

  for (const [name, value] of tags) {
    if (name === "EXT-X-TARGETDURATION") {
      targetDuration = value;
    }
    if (name === "EXT-X-ENDLIST") {
      endlist = true;
    }
    if (name === "EXT-X-PLAYLIST-TYPE") {
      playlistType = value;
    }
    if (name === "EXT-X-INDEPENDENT-SEGMENTS") {
      independentSegments = true;
    }
    if (name === "EXT-X-MEDIA-SEQUENCE") {
      mediaSequenceBase = value;
    }
    if (name === "EXT-X-DISCONTINUITY-SEQUENCE") {
      discontinuitySequenceBase = value;
    }
    if (name === "EXT-X-DATERANGE") {
      dateRanges.push(value);
    }
    if (name === "EXT-X-DEFINE") {
      defines.push(value);
    }
  }

  let segmentStart: number | null = null;
  tags.forEach(([name], index) => {
    if (segmentStart === null) {
      if (isSegmentTag(name)) {
        segmentStart = index;
      }
    } else if (name === "LITERAL") {
      const segmentTags = tags.slice(segmentStart, index + 1);
      const segment = parseSegment(segmentTags);
      segments.push(segment);
      segmentStart = null;
    }
  });

  normalizeSegments(segments);

  assert(targetDuration);

  return {
    targetDuration,
    endlist,
    playlistType,
    segments,
    independentSegments,
    mediaSequenceBase,
    discontinuitySequenceBase,
    dateRanges,
    defines,
  };
}

function parseSegment(tags: Tag[]): Segment {
  let duration: number | undefined;
  let discontinuity: boolean | undefined;
  let programDateTime: DateTime | undefined;
  let spliceInfo: SpliceInfo | undefined;
  let uri: string | undefined;
  let map: HLSMap | undefined;
  let key: HLSKey | undefined;

  for (const [name, value] of tags) {
    if (name === "EXTINF") {
      duration = value.duration;
    }
    if (name === "EXT-X-DISCONTINUITY") {
      discontinuity = true;
    }
    if (name === "EXT-X-PROGRAM-DATE-TIME") {
      programDateTime = value;
    }
    if (name === "EXT-X-CUE-IN") {
      spliceInfo = { type: "IN" };
    }
    if (name === "EXT-X-CUE-OUT") {
      spliceInfo = { type: "OUT", duration: value.duration };
    }
    if (name === "EXT-X-MAP") {
      map = value;
    }
    if (name === "EXT-X-KEY") {
      key = value;
    }
    if (name === "LITERAL") {
      uri = value;
    }
  }

  assert(uri, "parseSegment: uri not found");
  assert(duration, "parseSegment: duration not found");

  return {
    uri,
    duration,
    map,
    key,
    discontinuity,
    programDateTime,
    spliceInfo,
  };
}

function normalizeSegments(segments: Segment[]) {
  let lastMap: HLSMap | undefined;
  let lastKey: HLSKey | undefined;
  for (const segment of segments) {
    if (segment.map === undefined && lastMap) {
      segment.map = lastMap;
    } else {
      lastMap = segment.map;
    }

    if (segment.key === undefined && lastKey) {
      segment.key = lastKey;
    } else {
      lastKey = segment.key;
    }
  }
}

/**
 * Checks whether the tag name belongs to a segment.
 * @link https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis-16#section-4.4.4
 * @param name
 * @returns
 */
function isSegmentTag(name: TagName) {
  switch (name) {
    case "EXTINF":
    case "EXT-X-DISCONTINUITY":
    case "EXT-X-PROGRAM-DATE-TIME":
    case "EXT-X-MAP":
    case "EXT-X-CUE-OUT":
    case "EXT-X-CUE-IN":
    case "EXT-X-KEY":
      return true;
  }
  return false;
}
