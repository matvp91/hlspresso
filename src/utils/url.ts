import path from "node:path";

type ResolveUrlParams = {
  baseUrl: string;
  path: string;
};

export function resolveUrl(params: ResolveUrlParams) {
  if (params.path.startsWith("http://") || params.path.startsWith("https://")) {
    // Path appears to be the full url.
    return params.path;
  }

  const url = new URL(params.baseUrl);
  const urlPath = path.parse(url.pathname);

  // Rewrite the path.
  urlPath.base = params.path;
  url.pathname = path.format(urlPath);

  return url.toString();
}

export function replaceUrlParams(
  inputUrl: string,
  params?: Record<string, string | number | undefined>,
) {
  let url = inputUrl;
  const allParams = {
    ...params,
    // Default params defined below.
    random: Math.floor(Math.random() * 10_000),
  };

  const entries = Object.entries(allParams);
  for (const [key, value] of entries) {
    if (value === undefined) {
      continue;
    }
    url = url.replaceAll(`{${key}}`, value.toString());
  }

  return url;
}

export function getUrlCommonPrefix(urls: string[]) {
  if (!urls.length) {
    return "";
  }

  const value = urls.reduce((prefix, url) => {
    if (!prefix || !url) {
      return "";
    }
    let i = 0;
    const minLength = Math.min(prefix.length, url.length);
    while (i < minLength && prefix[i] === url[i]) {
      i++;
    }
    const lastSlash = prefix.lastIndexOf("/", i - 1);
    return lastSlash >= 0 ? prefix.slice(0, lastSlash + 1) : "";
  }, urls[0]);

  return value ?? "";
}
