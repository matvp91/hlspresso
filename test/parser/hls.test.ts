import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, test } from "vitest";
import { parseMainPlaylist } from "../../src/parser/hls";

const dirName = path.dirname(fileURLToPath(import.meta.url));

describe("Main playlist", () => {
  test.each([
    "Main_Basic.m3u8",
    "Main_Characteristics.m3u8",
    "Main_GroupedVariants.m3u8",
    "Main_IFrame.m3u8",
    "Main_AlternativeAudio.m3u8",
  ])("Parses %s", (name) => {
    const data = fs.readFileSync(`${dirName}/data/${name}`, "utf8");
    expect(parseMainPlaylist(data)).toMatchSnapshot();
  });
});
