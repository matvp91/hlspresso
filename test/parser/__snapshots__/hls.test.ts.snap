// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Main playlist > Parses Main_AlternativeAudio.m3u8 1`] = `
{
  "defines": [],
  "independentSegments": false,
  "medias": [
    {
      "autoSelect": true,
      "channels": undefined,
      "characteristics": undefined,
      "default": true,
      "groupId": "aac",
      "language": "en",
      "name": "English",
      "type": "AUDIO",
      "uri": "main/english-audio.m3u8",
    },
    {
      "autoSelect": true,
      "channels": undefined,
      "characteristics": undefined,
      "default": false,
      "groupId": "aac",
      "language": "de",
      "name": "Deutsch",
      "type": "AUDIO",
      "uri": "main/german-audio.m3u8",
    },
    {
      "autoSelect": false,
      "channels": undefined,
      "characteristics": undefined,
      "default": false,
      "groupId": "aac",
      "language": "en",
      "name": "Commentary",
      "type": "AUDIO",
      "uri": "commentary/audio-only.m3u8",
    },
  ],
  "variants": [
    {
      "audio": "aac",
      "bandwidth": 1280000,
      "codecs": "mp4a.40.2",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "low/video-only.m3u8",
    },
    {
      "audio": "aac",
      "bandwidth": 2560000,
      "codecs": "mp4a.40.2",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "mid/video-only.m3u8",
    },
    {
      "audio": "aac",
      "bandwidth": 7680000,
      "codecs": "mp4a.40.2",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "hi/video-only.m3u8",
    },
    {
      "audio": "aac",
      "bandwidth": 65000,
      "codecs": "mp4a.40.5",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "main/english-audio.m3u8",
    },
  ],
}
`;

exports[`Main playlist > Parses Main_Basic.m3u8 1`] = `
{
  "defines": [],
  "independentSegments": false,
  "medias": [],
  "variants": [
    {
      "audio": undefined,
      "bandwidth": 1280000,
      "codecs": "avc1.640029,mp4a.40.2",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "http://example.com/low.m3u8",
    },
    {
      "audio": undefined,
      "bandwidth": 2560000,
      "codecs": "avc1.640029,mp4a.40.2",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "http://example.com/mid.m3u8",
    },
    {
      "audio": undefined,
      "bandwidth": 7680000,
      "codecs": "avc1.640029,mp4a.40.2",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "http://example.com/hi.m3u8",
    },
    {
      "audio": undefined,
      "bandwidth": 65000,
      "codecs": "mp4a.40.5",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "http://example.com/audio-only.m3u8",
    },
  ],
}
`;

exports[`Main playlist > Parses Main_Characteristics.m3u8 1`] = `
{
  "defines": [],
  "independentSegments": false,
  "medias": [
    {
      "autoSelect": true,
      "channels": undefined,
      "characteristics": "public.accessibility.transcribes-spoken-dialog,public.easy-to-read",
      "default": true,
      "groupId": "aac",
      "language": "en",
      "name": "English",
      "type": "AUDIO",
      "uri": "main/english-audio.m3u8",
    },
    {
      "autoSelect": true,
      "channels": undefined,
      "characteristics": "public.accessibility.transcribes-spoken-dialog,public.easy-to-read",
      "default": false,
      "groupId": "aac",
      "language": "de",
      "name": "Deutsch",
      "type": "AUDIO",
      "uri": "main/german-audio.m3u8",
    },
    {
      "autoSelect": false,
      "channels": undefined,
      "characteristics": "public.accessibility.transcribes-spoken-dialog,public.easy-to-read",
      "default": false,
      "groupId": "aac",
      "language": "en",
      "name": "Commentary",
      "type": "AUDIO",
      "uri": "commentary/audio-only.m3u8",
    },
  ],
  "variants": [
    {
      "audio": "aac",
      "bandwidth": 1280000,
      "codecs": "mp4a.40.2",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "low/video-only.m3u8",
    },
    {
      "audio": "aac",
      "bandwidth": 2560000,
      "codecs": "mp4a.40.2",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "mid/video-only.m3u8",
    },
    {
      "audio": "aac",
      "bandwidth": 7680000,
      "codecs": "mp4a.40.2",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "hi/video-only.m3u8",
    },
    {
      "audio": "aac",
      "bandwidth": 65000,
      "codecs": "mp4a.40.5",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "main/english-audio.m3u8",
    },
  ],
}
`;

exports[`Main playlist > Parses Main_GroupedVariants.m3u8 1`] = `
{
  "defines": [],
  "independentSegments": true,
  "medias": [
    {
      "autoSelect": undefined,
      "channels": undefined,
      "characteristics": undefined,
      "default": true,
      "groupId": "aac_high",
      "language": undefined,
      "name": "English",
      "type": "AUDIO",
      "uri": "aac_high_eng.m3u8",
    },
    {
      "autoSelect": undefined,
      "channels": undefined,
      "characteristics": undefined,
      "default": false,
      "groupId": "aac_high",
      "language": undefined,
      "name": "Japanese",
      "type": "AUDIO",
      "uri": "aac_high_jp.m3u8",
    },
    {
      "autoSelect": undefined,
      "channels": undefined,
      "characteristics": undefined,
      "default": true,
      "groupId": "aac_mid",
      "language": undefined,
      "name": "English",
      "type": "AUDIO",
      "uri": "aac_mid_eng.m3u8",
    },
    {
      "autoSelect": undefined,
      "channels": undefined,
      "characteristics": undefined,
      "default": false,
      "groupId": "aac_mid",
      "language": undefined,
      "name": "Japanese",
      "type": "AUDIO",
      "uri": "aac_mid_jp.m3u8",
    },
    {
      "autoSelect": undefined,
      "channels": undefined,
      "characteristics": undefined,
      "default": true,
      "groupId": "aac_low",
      "language": undefined,
      "name": "English",
      "type": "AUDIO",
      "uri": "aac_low_eng.m3u8",
    },
    {
      "autoSelect": undefined,
      "channels": undefined,
      "characteristics": undefined,
      "default": false,
      "groupId": "aac_low",
      "language": undefined,
      "name": "Japanese",
      "type": "AUDIO",
      "uri": "aac_low_jp.m3u8",
    },
  ],
  "variants": [
    {
      "audio": "aac_high",
      "bandwidth": 6000000,
      "codecs": undefined,
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "1080p.m3u8",
    },
    {
      "audio": "aac_mid",
      "bandwidth": 3000000,
      "codecs": undefined,
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "720p.m3u8",
    },
    {
      "audio": "aac_mid",
      "bandwidth": 1500000,
      "codecs": undefined,
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "540p.m3u8",
    },
    {
      "audio": "aac_low",
      "bandwidth": 1000000,
      "codecs": undefined,
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "360p.m3u8",
    },
  ],
}
`;

exports[`Main playlist > Parses Main_IFrame.m3u8 1`] = `
{
  "defines": [],
  "independentSegments": false,
  "medias": [],
  "variants": [
    {
      "audio": undefined,
      "bandwidth": 1280000,
      "codecs": "avc1.640029,mp4a.40.2",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "low/audio-video.m3u8",
    },
    {
      "audio": undefined,
      "bandwidth": 2560000,
      "codecs": "avc1.640029,mp4a.40.2",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "mid/audio-video.m3u8",
    },
    {
      "audio": undefined,
      "bandwidth": 7680000,
      "codecs": "avc1.640029,mp4a.40.2",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "hi/audio-video.m3u8",
    },
    {
      "audio": undefined,
      "bandwidth": 65000,
      "codecs": "mp4a.40.5",
      "resolution": undefined,
      "subtitles": undefined,
      "uri": "audio-only.m3u8",
    },
  ],
}
`;
